ui/lib/dialogs/add_restriction_rule_dialog.dart

```diff
@@ -0,0 +1,163 @@
+import 'package:flutter/material.dart';
+import 'package:ui/common_widgets/error_message_widget.dart';
+import 'package:ui/common_widgets/text_field_list.dart';
+import 'package:ui/locator.dart';
+import 'package:ui/services/fraud_detect_service.dart';
+import 'package:ui/services/proto/fraud-detection-service.pb.dart';
+
+class AddRestrictionRuleDialog extends StatefulWidget {
+  const AddRestrictionRuleDialog({super.key});
+
+  @override
+  State<AddRestrictionRuleDialog> createState() =>
+      _AddRestrictionRuleDialogState();
+}
+
+class _AddRestrictionRuleDialogState extends State<AddRestrictionRuleDialog> {
+  late final TextEditingController idController;
+  late final TextEditingController ruleNameController;
+  late final TextEditingController predicateController;
+
+  bool isLoading = false;
+  String errorMessage = "";
+
+  @override
+  void initState() {
+    idController = TextEditingController();
+    ruleNameController = TextEditingController();
+    predicateController = TextEditingController();
+    super.initState();
+  }
+
+  @override
+  void dispose() {
+    idController.dispose();
+    ruleNameController.dispose();
+    predicateController.dispose();
+    super.dispose();
+  }
+
+  void _onCancel(BuildContext context) {
+    Navigator.of(context).pop();
+  }
+
+  Future<void> _onSubmit(BuildContext context) async {
+    //TODO: 3lab content use form validation
+    if (idController.text.isEmpty ||
+        ruleNameController.text.isEmpty ||
+        predicateController.text.isEmpty) {
+      setState(() => errorMessage = "No field can be empty");
+      return;
+    }
+
+    setState(() {
+      isLoading = true;
+      errorMessage = "";
+    });
+
+    final service = locator.get<FraudDetectService>();
+
+    //check if id is a number
+    if (int.tryParse(idController.text.trim()) == null) {
+      setState(() {
+        isLoading = false;
+        errorMessage = "Id should be a number";
+      });
+      return;
+    }
+
+    //TODO: move rules State to local page Provider so i don't have to send an extra request
+    //send request for a list of rules
+    late final RestrictionRules rules;
+    try {
+      rules = await service.restrictionRulesServiceClient
+          .getRestrictionRules(Empty());
+    } catch (e) {
+      setState(() {
+        isLoading = false;
+        errorMessage = "Failed to receive data";
+      });
+      return;
+    }
+
+    //Check if a row with this id already exists
+    if (rules.rule.map((x) => x.id).contains(idController.text.trim())) {
+      setState(() {
+        isLoading = false;
+        errorMessage = "A row with this id already exists";
+      });
+      return;
+    }
+
+    //add new rule
+    final idString = idController.text.trim();
```

Comment:
As discussed privately, there is no need to add ID field to form when create new rule. We can just generate UUID (see https://pub.dev/packages/uuid).